torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1528,249 @@ def __exit__(self, a, b, c):
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
+
+        key_values = (
+            func,
+            # Translate any FakeTensor args to metadata.
+            self._translate_arg(args),
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
+            # mode can introduce a DynamicOutputShapeException where it wasn't seen on
+            # a previous instance of the same func and arg combination.
+            self.shape_env and self.shape_env.allow_dynamic_output_shape_ops,
+        )
+        return _DispatchCacheKey(key_values)
+
+    def _translate_arg(self, arg: Any) -> Any:
+        """
+        Translate the provided arg into a form suitable for caching at FakeTensor
+        dispatch, i.e., convert unhashable types like lists & dicts into tuples and
+        convert FakeTensors into metadata. Raises _BypassDispatchCache to signal
+        unsupported cases that should bypass caching.
+        """
+        if isinstance(arg, (list, tuple)):
+            return tuple(self._translate_arg(x) for x in arg)
+        elif isinstance(arg, dict):
+            return tuple((k, self._translate_arg(v)) for k, v in arg.items())
+        elif isinstance(arg, (torch.SymBool, torch.SymInt, torch.SymFloat)):
+            raise _BypassDispatchCache("symbolic shape")
+        elif isinstance(arg, FakeTensor):
+            if not self.is_our_fake(arg):
+                raise _BypassDispatchCache("not our fake")
+            if arg._has_symbolic_sizes_strides:
+                raise _BypassDispatchCache("symbolic shape")
```

Comment:
Is there a plan for dynamic shapes? (just support static for now and worry / think about dyamic shape plan later?)

Reply:
> (just support static for now and worry / think about dyamic shape plan later?)

Yeah, that's the plan. Supporting dynamic shapes for the code cache was a bit tricky. Basically, we had to:
1) create a guards expression when we encounter dynamic shapes
2) save that expression with the cache entry
3) on a potential "hit", evaluate the expression in the current context to see if a "hit" is really a "hit"

I presume we'd have to take a similar approach to support dynamic shapes here? But I think it's trickier here because the performance of the cache lookup is more critical compared to the code cache.